Generics support

majakusber opened the pull request:
Revert of PR #1411, adding fixes for parsing edge cases of generics, adding unit tests

Commit:
Revert "Revert "TG-374 Feature/java support generics""

This reverts commit 3b3fee34be5bdbd119f6cdb90b6bdfe6ba4ad4d0.

Commit:
Fixing handling the case of *

Commit:
Classes that aren't generic but inherit from a generic type have a signature

This is a lazy fix to stop trying to create java_generics_class_typet
when the signature doesn't have the correct format

Commit:
When dealing with generic arrays we should treat them like ref arrays too

For some reason, A indicates a reference array

Commit:
Correctly handle nested generic types

Previously looked for first > and ;, but in nested types we need to find
the last semi-colon and the matching angle bracket.

Commit:
Adding conversions for wild cards

Need to check all these are actually required after the other fixes

Commit:
Adding some useful debug info

Commit:
Correcting a typo

Commit:
Deal with generic methods

The signature of a method can start with an angle bracket if it is a
generic method on a non-generic type or a static generic method.
Previously this would cause us to return a nil type rather than a code
type. This information is currently ignored and instead we just skip
over the Formal Type parameters and proceed using normal type
evaluation.

Commit:
Extending the tests for the generic class

Commit:
Adding unit tests for parsing wild card functions

Commit:
Correct handling of the java generic class signature

Previously the signature parsing would be confused into thinking that
there were two generic parameters when there were two bounds on the
first parameter. There exists TG-674 to make the conversion work
correctly.

Commit:
Adding unit test for java class that inherits from a generic class

Commit:
Applied use of utility function for loading a class file

Commit:
Adding tests for generic functions

Commit:
Adding unit test for generic array

Commit:
Adding unit test for recursive generic class

Commit:
Cleaning java files

thk123 commented on regression/cbmc-java/enum1/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
These hopefully don't need to be turned off any more as it shouldn't crash when they are found.

thk123 commented:
I suggest reviewers look at [this diff](https://github.com/diffblue/cbmc/pull/1434/files/e36d2cd25d121654c22df8c097ad10b017245e45..9b4662d3bc41db55bcca7948b726bcb1b295fd3c) as the original PR (#1322) is reapplied whole sale and has already been reviewed

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -94,7 +94,7 @@ void java_bytecode_convert_classt::convert(const classt &c)
 
   java_class_typet class_type;
-  if(c.signature.has_value())
+  if(c.signature.has_value() && c.signature.value()[0]=='<')
```
@mgudemann This was to fix parsing classes that extend from generic classes

thk123 commented on src/java_bytecode/java_types.cpp:
```diff
@@ -261,7 +285,8 @@ typet java_type_from_string(
       if(subtype_letter=='L' || // [L denotes a reference array of some sort.
-         subtype_letter=='[')   // Array-of-arrays
+         subtype_letter=='[' || // Array-of-arrays
+         subtype_letter=='T')   // Array of generic types
```
@mgudemann Deals with reference types that are of the form `T []`

thk123 commented on src/java_bytecode/java_types.cpp:
```diff
@@ -389,6 +414,55 @@ typet java_type_from_string(
       return java_reference_type(symbol_type);
     }
+  case '*':
+  {
```
Probably these three cases (`'*'`, `'-'`, `'+'`) should be handled by throwing an exception rather than this ugly partial handling.

chrisr-diffblue replied:
I'd agree with @thk123

thk123 replied:
For the record - I should say it was me who wrote the  "ugly partial handling" :stuck_out_tongue:

thk123 commented on src/java_bytecode/java_types.cpp:
```diff
@@ -438,13 +512,16 @@ std::vector<typet> java_generic_type_from_string(
     "Class signature has unexpected format");
   std::string signature(src.substr(1, signature_end-1));
 
+  if(signature.find(";:")!=std::string::npos)
```
@mgudemann this is to resolve when there are multiple bounds on a type

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/Bar.java:
```diff
@@ -0,0 +1,2 @@
+class Bar extends Foo
+{}
```
Nit: this file (and others) are missing a new line at the end of the file (and others). You can make sure CLion inserts these using Settings > Editor > General > Ensure Line Feed at end of file on save

thk123 reviewed (commented):
Some small nits - I won't approve as I had a hand in writing some of this.

chrisr-diffblue commented on src/java_bytecode/java_types.cpp:
```diff
@@ -389,6 +414,55 @@ typet java_type_from_string(
+    PRECONDITION(!class_name.empty());
+//    INVARIANT(
+//      src.substr(1, src.size()-2)=="*",
+//      "Wildcard types do not have a name");
```
I assume these commented out INVARIANT's are intended to be re-instated once the generics support improves - if so, is it worth marking them all with FIXME or TODO comments as well to make that clearer? @thk123 is there any common practice here to include a link to a Jira or PR for in the comment to indicate what the work is covered under?

chrisr-diffblue reviewed (commented):
Looks pretty good, but like @thk123 a couple if minor nits. Happy to approve if they are addressed, or there's a good explanation why the PR is fine as it is.

mgudemann commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -0,0 +1,54 @@
+  // }
+
+  // Test a wild card generic bound by a class
+  // public static void processBoundClassGenericDoubleBound(SimpleGeneric<? extends Foo & BasicInterface> x) {
```
if wildcards can have multiplle bounds, we should have this as test

thk123 replied:
Sorry comment not clear - wildcards can not have multiple bounds - this java code doesn't compile :)

mgudemann commented on unit/java_bytecode/java_bytecode_parse_generics/GenericFunctions.java:
```diff
@@ -0,0 +1,20 @@
+    assert(x.t.getX() == 4);
+  }
+
+  public static <T extends Foo & BasicInterface> void processDoubleUpperBoundClassGeneric(SimpleGeneric<T> x) {
```
maybe add test with two interface bounds and no class, too

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -158,13 +167,54 @@ exprt java_bytecode_promotion(const exprt &expr)
+      // The method signature can optionally have a collection of formal
+      // type parameters (e.g. on generic methods on non-generic classes
+      // or generic static methods). For now we skip over this part of the
+      // signature and continue parsing the rest of the signature as normal
```
maybe give a concrete example here what would be skipped
could this '<' case and the '<' case inside 'L' - object name parsing be merged ?

thk123 replied:
No - this should only occur for functions that look like:

```java
void <T> foo(T t)
```

etc (i.e. functions that are generic as opposed to part of a generic class). Perhaps a more robust way of dealing with this is to throw and exception and fall back to the descriptor in this case (since this is broadly what this is doing anyway)

thk123 replied:
Proposed comment extension:
```C++
// So for example, the following java method:
// static void <T, U> foo(T t, U u, int x);
// Would have a signature that looks like:
// <T:Ljava/lang/Object;U:Ljava/lang/Object;>(TT;TU;I)V
// So we skip all inside the angle brackets and parse the rest of the string:
// (TT;TU;I)V
```

majakusber replied:
Thank you Thomas, adding it in..

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -233,14 +302,109 @@ typet java_type_from_string(const std::string &src)
       // ends on ;
       if(src[src.size()-1]!=';')
         return nil_typet();
       std::string class_name=src.substr(1, src.size()-2);
```
as discussed, this shadows the `class_name` parameter and should be moved down to where this string is used (after the generic signature case)

mgudemann commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -93,6 +94,33 @@ void java_bytecode_convert_classt::convert(const classt &c)
+    }
+    catch(unsupported_java_class_siganture_exceptiont)
+    {
+      // Do nothing: we don't support parsing for example double bounded or
```
maybe here we should fall back to using the descriptor, and emit a warning, in order to degrade gracefully

thk123 replied:
My understanding is the `descriptor` won't have any generic type information so isn't applicable here (We are only trying to fill out generic information for the class)

thk123 replied:
Adding a warning is a good idea however.

thk123 reviewed (changes requested):
Blocking until we've resolved issues with the TG repo

Commit:
Reverting method descriptor loading

The signature and the descriptor of a method do not match up (see java
spec: The signature and descriptor (§4.3.3) of a given method or
constructor may not correspond exactly, due to compiler-generated
artifacts. In particular, the number of TypeSignatures that encode
formal arguments in MethodTypeSignature may be less than the number of
ParameterDescriptors in MethodDescriptor.) Reverting this to descriptor
solves the current problem with passing tests.

Commit:
Disabling part of the unit test for generic classes

Due to methods no longer being recognised as generics
(signature/descriptor issue).

Commit:
Turning back on regression tests

Commit:
Handling wild card generics with exception

Instead of partial handling from before.

Commit:
Adding a warning and a commentary for unsupported generics, cleaning

majakusber commented:
Thank you everyone for the reviews, I've now pushed all the changes according to your notes. Let me know what you think.

thk123 commented:
Issues with TG repo have been resolved: just need this PRs conflict resolved and diffblue/test-gen#1080 to pass.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -0,0 +1,323 @@
+
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
```
This test case doesn't actually test what is stated in the 'THEN' clause - and I think this functionality has broken. I've just rebased my work on top of this PR, and one of my unit tests is failing because of this. Previously, you could call `is_java_generic_parameter(code.return_type())` and get the answer `true` for this function. I'm not finding this is returning `false` instead, and doing 'method_symbol.show(std::cout)` is showing that the return type for this method is not marked as being generic in any way.

majakusber replied:
Thanks for this, this has also been failing the CI tests. The reason is that the return type of the lookup method for symbol table has changed (commit #1444 ) - now it's a reference wrapper around symbolt instead of symbolt.

chrisr-diffblue replied:
Yes, that is part of the failure (it's broken my work too :-) ), but my broader point is that this test case doesn't *actually* test what it claims to test - it claims to test that it parses methods with a generic return type, but it only does that partially - it checks that it can parse the method, and that the symbol for the method exists, and is an ID_code object - but it *doesn't* check that the return type of that `code_typet` actually records that it is generic. So, for instance, we could be parsing the the method and then throwing away all the generic information and this test as it is currently written would still be passing. This appears to have happened at some point, because when my code was based off @thk123 branch, the `is_java_generic_parameter(code.return_type())` was correctly returning true for this case, whereas with this branch I'm seeing that the generic information is no longer stored for the return type.

thk123 replied:
Yes - this is the mess with descriptors of methods being used. As discussed I think before we do [TG-865](https://diffblue.atlassian.net/browse/TG-865) it won't work. The THEN should be updated in a similar way to the TODO directly below.

NlightNFotis commented on src/java_bytecode/java_types.cpp:
```diff
@@ -158,13 +167,61 @@ exprt java_bytecode_promotion(const exprt &expr)
+      // So we skip all inside the angle brackets and parse the rest of the
+      // string:
+      // (TT;TU;I)V
+      size_t closing_generic=find_closing_delimiter(src, 0, '<', '>');
```
There is a particular problem with this function when passed a `src` string with length `1` only, and it results in a huge recursive call stack, and makes the program result in a `Segmentation Violation` fault. I shall send tests and perhaps a fix in a bit.

Commit:
Resolving name shadowing

class_name parameter renamed to class_name_prefix

mgudemann reviewed (approved):
LGTM, please squash some of the commits before merge

Commit:
Updating the calls of lookup method to reflect the new return type

The return type of the method lookup for the symbol table has changed in
a recent CBMC commit.

Commit:
Disabling part of the unit test for generic classes

Reintroduce when descriptor/signature issue is resolved.

Commit:
Adding exception and tests for missing closing delimiter

Commit:
Removing signature parsing for local variables

Might need changing once the signature/descriptor issue is resolved.
